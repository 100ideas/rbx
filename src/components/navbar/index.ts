import "./navbar.sass";
export { default } from "./Navbar";
