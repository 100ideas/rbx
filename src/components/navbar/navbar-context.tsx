import React from "react";

import { noop } from "../../utils";

export interface NavbarContextState {
  active: boolean;
  setActive: (value: boolean) => void;
}

export const initialState: NavbarContextState = {
  active: false,
  setActive: noop,
};

export const NavbarContext = React.createContext<NavbarContextState>(
  initialState,
);
