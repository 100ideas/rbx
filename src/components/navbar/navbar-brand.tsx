import { cx } from "emotion";
import React from "react";

import { Element } from "@/components/element";
import { ModifierProps } from "@/modifiers";

export type NavbarBrandProps = Prefer<
  ModifierProps,
  React.HTMLAttributes<HTMLDivElement>
>;

export const NavbarBrand = React.forwardRef<HTMLDivElement, NavbarBrandProps>(
  (props, ref) => {
    const { className, ...rest } = props;
    return (
      <Element {...rest} ref={ref} className={cx("navbar-brand", className)} />
    );
  },
);
