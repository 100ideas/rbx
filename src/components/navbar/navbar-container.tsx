import { cx } from "emotion";
import React from "react";

import { asExoticComponent } from "@/components/exotic";
import { ModifierProps, modify } from "@/modifiers";

export type NavbarContainerModifierProps = Partial<{
  position: "start" | "end";
}>;

export type NavbarContainerProps = ModifierProps & NavbarContainerModifierProps;

export const NavbarContainer = asExoticComponent<NavbarContainerProps, "div">(
  (props, ref) => {
    const { as, position, ...rest } = modify(props);
    rest.className = cx(rest.className, {
      [`navbar-${position}`]: position,
    });
    return React.createElement(as!, { ref, ...rest });
  },
  "div",
);

NavbarContainer.defaultProps = Object.assign(
  { position: "start" },
  NavbarContainer.defaultProps,
);
