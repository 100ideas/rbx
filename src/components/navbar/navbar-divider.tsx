import { cx } from "emotion";
import React from "react";

import { Element } from "@/components/element";
import { ModifierProps } from "@/modifiers";

export type NavbarDividerProps = Prefer<
  ModifierProps,
  React.HTMLAttributes<HTMLDivElement>
>;

export const NavbarDivider = React.forwardRef<
  HTMLDivElement,
  NavbarDividerProps
>(({ className, ...props }, ref) => (
  <Element {...props} ref={ref} className={cx("navbar-divider", className)} />
));
