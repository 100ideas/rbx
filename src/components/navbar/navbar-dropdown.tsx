import { cx } from "emotion";
import React from "react";

import { asExoticComponent } from "@/components/exotic";
import { ModifierProps, modify } from "@/modifiers";

export type NavbarDropdownModifierProps = Partial<{
  boxed: boolean;
  right: boolean;
}>;

export type NavbarDropdownProps = ModifierProps & NavbarDropdownModifierProps;

export const NavbarDropdown = asExoticComponent<NavbarDropdownProps, "span">(
  (props, ref) => {
    const { as, boxed, right, ...rest } = modify(props);
    rest.className = cx("navbar-dropdown", rest.className, {
      "is-boxed": boxed,
      "is-right": right,
    });
    return React.createElement(as!, { ref, ...rest });
  },
  "span",
);

NavbarDropdown.defaultProps = Object.assign(
  {
    boxed: false,
    right: false,
  },
  NavbarDropdown.defaultProps,
);
