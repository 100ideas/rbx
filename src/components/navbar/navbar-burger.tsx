import { cx } from "emotion";
import React from "react";

import { Element } from "@/components/element";
import { ModifierProps, modify } from "@/modifiers";
import { NavbarContext } from "./navbar-context";

export type NavbarBurgerProps = Prefer<
  ModifierProps,
  React.HTMLAttributes<HTMLDivElement>
>;

export const NavbarBurger = React.forwardRef<HTMLDivElement, NavbarBurgerProps>(
  (props, ref) => {
    const { className, style, onClick, ...rest } = modify(props);
    return (
      <NavbarContext.Consumer>
        {({ active, setActive }) => (
          <Element
            className={cx("navbar-burger", className, {
              "is-active": active,
            })}
            onClick={() => setActive(!active)}
            ref={ref}
            role="button"
            style={{ outline: "none", ...style }}
            tabIndex={0}
            {...rest}
          >
            <span />
            <span />
            <span />
          </Element>
        )}
      </NavbarContext.Consumer>
    );
  },
);
