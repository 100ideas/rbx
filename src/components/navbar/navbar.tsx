import { cx } from "emotion";
import React from "react";

import { Element } from "@/components/element";
import { asExoticComponent } from "@/components/exotic";
import { ModifierProps, modify } from "@/modifiers";
import { Colors } from "@/modifiers/color";
import { canUseDOM } from "@/utils";
import { NavbarBrand } from "./navbar-brand";
import { NavbarBurger } from "./navbar-burger";
import { NavbarContainer } from "./navbar-container";
import { NavbarContext } from "./navbar-context";
import { NavbarDivider } from "./navbar-divider";
import { NavbarDropdown } from "./navbar-dropdown";
import { NavbarItem } from "./navbar-item";
import { NavbarLink } from "./navbar-link";
import { NavbarMenu } from "./navbar-menu";

let htmlClass = "";

export const getHtmlClasses = () => htmlClass;

export type NavbarModifierProps = Partial<{
  active: boolean;
  children: React.ReactNode;
  className: string;
  color: Colors;
  fixed: "top" | "bottom";
  style: React.CSSProperties;
  transparent: boolean;
}>;

export type NavbarProps = ModifierProps & NavbarModifierProps;

type NavbarControllerProps = NavbarProps & {
  innerRef: React.Ref<HTMLDivElement>;
};

interface NavbarControllerState {
  active: boolean;
}

class NavbarController extends React.PureComponent<
  NavbarControllerProps,
  NavbarControllerState
> {
  public static defaultProps = {
    active: false,
    children: null,
    transparent: false,
  };

  public readonly state: NavbarControllerState;

  constructor(props: NavbarControllerProps) {
    super(props);
    this.state = { active: props.active! };
  }

  public setActive = (value: boolean) => {
    this.setState({ active: value });
  }

  public componentWillUnmount() {
    if (canUseDOM()) {
      const { fixed } = this.props;
      const html = window.document.querySelector("html");
      if (html) {
        html.classList.remove(`has-navbar-fixed-${fixed}`);
      }
    }
  }

  public render() {
    this.manageHtmlAttributes();

    const {
      active, // only used for initialState (in constructor)
      color,
      fixed,
      innerRef,
      transparent,
      ...rest
    } = modify(this.props);

    rest.className = cx("navbar", rest.className, {
      "is-transparent": transparent,
      [`is-fixed-${fixed}`]: fixed,
      [`is-${color}`]: color,
    });

    return (
      <NavbarContext.Provider
        value={{ active: this.state.active, setActive: this.setActive }}
      >
        <Element ref={innerRef} role="navigation" {...rest} />
      </NavbarContext.Provider>
    );
  }

  private manageHtmlAttributes() {
    if (canUseDOM()) {
      const html = window.document.querySelector("html");
      if (html) {
        html.classList.remove("has-navbar-fixed-top");
        html.classList.remove("has-navbar-fixed-bottom");
        if (this.props.fixed) {
          htmlClass = `has-navbar-fixed-${this.props.fixed}`;
          html.classList.add(`has-navbar-fixed-${this.props.fixed}`);
        }
      }
    }
  }
}

export const Navbar = Object.assign(
  asExoticComponent<NavbarProps, "nav">(
    (props, ref) => <NavbarController innerRef={ref} {...props} />,
    "nav",
  ),
  {
    Brand: NavbarBrand,
    Burger: NavbarBurger,
    Container: NavbarContainer,
    Divider: NavbarDivider,
    Dropdown: NavbarDropdown,
    Item: NavbarItem,
    Link: NavbarLink,
    Menu: NavbarMenu,
  },
);
