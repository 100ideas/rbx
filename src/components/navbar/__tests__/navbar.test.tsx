import React from "react";
import { renderToString } from "react-dom/server";
import renderer, { ReactTestRenderer } from "react-test-renderer";

import { setupWindow, teardownWindow } from "@/__tests__/helpers";
import { getHtmlClasses, Navbar } from "../navbar";

describe("Navbar component", () => {
  let component: ReactTestRenderer;

  beforeEach(() => {
    setupWindow();
  });

  afterEach(() => {
    if (component && component.unmount) {
      component.unmount();
    }
    teardownWindow();
  });

  it("should Exist", () => {
    expect(Navbar).toMatchSnapshot();
  });

  it("should render on server side", () => {
    const html = renderToString(<Navbar fixed="top" />);
    expect(html).toMatchSnapshot();
    expect(getHtmlClasses()).toMatchSnapshot("top");
  });

  it("should have Navbar classname", () => {
    component = renderer.create(
      <Navbar>
        <Navbar.Brand>
          <Navbar.Item<"a"> as="a" href="#">
            <img src="/" alt="" />
          </Navbar.Item>
          <Navbar.Burger className="trigger-menu" />
        </Navbar.Brand>
        <div>TEST</div>
        <Navbar.Menu>
          <Navbar.Container>
            <Navbar.Item dropdown hoverable>
              <Navbar.Link>Docs</Navbar.Link>
              <Navbar.Dropdown boxed>
                <Navbar.Item href="#">Home</Navbar.Item>
                <Navbar.Divider />
                <Navbar.Item href="#">Home</Navbar.Item>
              </Navbar.Dropdown>
            </Navbar.Item>
          </Navbar.Container>
        </Navbar.Menu>
      </Navbar>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should concat Bulma class with classes in props", () => {
    component = renderer.create(<Navbar className="other-class test" />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should render as an html section", () => {
    component = renderer.create(<Navbar<"section"> as="section" />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should have custom inline styles", () => {
    component = renderer.create(<Navbar style={{ width: 200, zIndex: 1 }} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should be fixed on top", () => {
    component = renderer.create(<Navbar fixed="top" />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
