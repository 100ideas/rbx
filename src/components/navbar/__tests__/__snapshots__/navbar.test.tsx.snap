// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar component should Exist 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "Brand": Object {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "Burger": Object {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "Container": Object {
    "$$typeof": Symbol(react.forward_ref),
    "defaultProps": Object {
      "as": "div",
      "position": "start",
    },
    "render": [Function],
  },
  "Divider": Object {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "Dropdown": Object {
    "$$typeof": Symbol(react.forward_ref),
    "defaultProps": Object {
      "as": "span",
      "boxed": false,
      "right": false,
    },
    "render": [Function],
  },
  "Item": Object {
    "$$typeof": Symbol(react.forward_ref),
    "defaultProps": Object {
      "active": false,
      "as": "a",
      "dropdown": false,
      "dropdownUp": false,
      "hoverable": false,
    },
    "render": [Function],
  },
  "Link": Object {
    "$$typeof": Symbol(react.forward_ref),
    "defaultProps": Object {
      "as": "span",
    },
    "render": [Function],
  },
  "Menu": Object {
    "$$typeof": Symbol(react.forward_ref),
    "render": [Function],
  },
  "defaultProps": Object {
    "as": "nav",
  },
  "render": [Function],
}
`;

exports[`Navbar component should be fixed on top 1`] = `
<nav
  className="navbar is-fixed-top"
  role="navigation"
/>
`;

exports[`Navbar component should concat Bulma class with classes in props 1`] = `
<nav
  className="navbar other-class test"
  role="navigation"
/>
`;

exports[`Navbar component should have Navbar classname 1`] = `
<nav
  className="navbar"
  role="navigation"
>
  <div
    className="navbar-brand"
  >
    <a
      className="navbar-item"
      href="#"
    >
      <img
        alt=""
        src="/"
      />
    </a>
    <div
      className="navbar-burger trigger-menu"
      onClick={[Function]}
      role="button"
      style={
        Object {
          "outline": "none",
        }
      }
      tabIndex={0}
    >
      <span />
      <span />
      <span />
    </div>
  </div>
  <div>
    TEST
  </div>
  <div
    className="navbar-menu"
  >
    <div
      className="navbar-start"
    >
      <span
        className="navbar-item has-dropdown is-hoverable"
      >
        <span
          className="navbar-link"
        >
          Docs
        </span>
        <span
          className="navbar-dropdown is-boxed"
        >
          <a
            className="navbar-item"
            href="#"
          >
            Home
          </a>
          <div
            className="navbar-divider"
          />
          <a
            className="navbar-item"
            href="#"
          >
            Home
          </a>
        </span>
      </span>
    </div>
  </div>
</nav>
`;

exports[`Navbar component should have custom inline styles 1`] = `
<nav
  className="navbar"
  role="navigation"
  style={
    Object {
      "width": 200,
      "zIndex": 1,
    }
  }
/>
`;

exports[`Navbar component should render as an html section 1`] = `
<section
  className="navbar"
  role="navigation"
/>
`;

exports[`Navbar component should render on server side 1`] = `"<nav role=\\"navigation\\" class=\\"navbar is-fixed-top\\"></nav>"`;

exports[`Navbar component should render on server side: top 1`] = `"has-navbar-fixed-top"`;
